src/node.rs

```diff
@@ -1765,3 +1896,492 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
         not_implemented("fee history")
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{http_fork_source::HttpForkSource, node::InMemoryNode};
+    use zksync_types::{api::BlockNumber, Address, L2ChainId, Nonce, PackedEthSignature};
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_get_block_by_hash_for_produced_block() {
+        let node = InMemoryNode::<HttpForkSource>::default();
+
+        let private_key = H256::random();
+        let from_account = PackedEthSignature::address_from_private_key(&private_key)
+            .expect("failed generating address");
+        node.set_rich_account(from_account);
+        let mut tx = L2Tx::new_signed(
+            Address::random(),
+            vec![],
+            Nonce(0),
+            Fee {
+                gas_limit: U256::from(1_000_000),
+                max_fee_per_gas: U256::from(250_000_000),
+                max_priority_fee_per_gas: U256::from(250_000_000),
+                gas_per_pubdata_limit: U256::from(20000),
+            },
+            U256::from(1),
+            L2ChainId(260),
+            &private_key,
+            None,
+            Default::default(),
+        )
+        .unwrap();
+        tx.set_input(vec![], H256::repeat_byte(0x01));
+
+        node.apply_txs(vec![tx.into()]).expect("failed applying tx");
+
+        let expected_block_hash =
+            H256::from_str("0x89c0aa770eba1f187235bdad80de9c01fe81bca415d442ca892f087da56fa109")
+                .unwrap();
+        let actual_block = node
+            .get_block_by_hash(expected_block_hash, false)
+            .await
+            .expect("failed fetching block by hash")
+            .expect("no block");
+
+        assert_eq!(expected_block_hash, actual_block.hash);
+        assert_eq!(U64::from(1), actual_block.number);
+        assert_eq!(Some(U64::from(1)), actual_block.l1_batch_number);
+    }
+
+    #[tokio::test]
+    async fn test_get_block_by_hash_uses_fork_source_and_caches_it() {
+        let input_block_hash_str =
+            "0x0101010101010101010101010101010101010101010101010101010101010101";
+        let input_block_hash = H256::from_str(input_block_hash_str).expect("invalid hash");
+
+        let mock_server = crate::testing::MockServer::run();
+        let mock_block_number = 8;
+        mock_server.expect(serde_json::json!({
+            "jsonrpc": "2.0",
+            "id": 0,
+            "method": "eth_getBlockByHash",
+            "params": [
+                input_block_hash_str,
+                true
+            ],
+        }), serde_json::json!({
+            "jsonrpc": "2.0",
+            "id": 0,
+            "result": {
```

Comment:
this large result repeats in almost every test. Consider creating a local function that returns it instead.